Storage/SQL/SQLiteBookmarksModel.swift

```diff
@@ -0,0 +1,832 @@
+        ", keyword = NULL" +
+        ", description = NULL" +
+        ", parentid = NULL" +
+        ", parentName = NULL" +
+        ", folderName = NULL" +
+        ", queryId = NULL" +
+        " WHERE \(whereFragment) AND sync_status = \(SyncStatus.Changed.rawValue)"
+
+        return sql
+    }
+    /**
+     * This depends on the record's parent already being overridden if necessary.
+     */
+    private func removeLocalByGUID(guid: GUID) -> Success {
+        let args: Args = [guid]
+
+        // Find the index we're currently occupying.
+        let previousIndexSubquery = "SELECT idx FROM \(TableBookmarksLocalStructure) WHERE child = ?"
+
+        // Fix up the indices of subsequent siblings.
+        let updateIndices =
+        "UPDATE \(TableBookmarksLocalStructure) SET idx = (idx - 1) WHERE idx > (\(previousIndexSubquery))"
+
+        // If the bookmark is New, delete it outright.
+        let deleteNew =
+        "DELETE FROM \(TableBookmarksLocal) WHERE guid = ? AND sync_status = \(SyncStatus.New.rawValue)"
+
+        // If the bookmark is Changed, mark it as deleted and bump its modified time.
+        let markChanged = self.getMarkDeletedSQLWithWhereFragment("guid = ?")
+
+        // Its parent must be either New or Changed, so we don't need to re-mirror it.
+        // TODO: bump the parent's modified time, because the child list changed?
+
+        // Now delete from structure.
+        let deleteStructure =
+        "DELETE FROM \(TableBookmarksLocalStructure) WHERE child = ?"
+
+        return self.db.run([
+            (updateIndices, args),
+            (deleteNew, args),
+            (markChanged, args),
+            (deleteStructure, args),
+        ])
+    }
+}
+
+class SQLiteBookmarkFolder: BookmarkFolder {
+    private let cursor: Cursor<BookmarkNode>
+    override var count: Int {
+        return cursor.count
+    }
+
+    override subscript(index: Int) -> BookmarkNode {
+        let bookmark = cursor[index]
+        return bookmark! as BookmarkNode
+    }
+
+    init(guid: String, title: String, children: Cursor<BookmarkNode>) {
+        self.cursor = children
+        super.init(guid: guid, title: title)
+    }
+}
+
+class BookmarkFactory {
+    private class func addIcon(bookmark: BookmarkNode, row: SDRow) {
+        // TODO: share this logic with SQLiteHistory.
+        if let faviconURL = row["iconURL"] as? String,
+           let date = row["iconDate"] as? Double,
+           let faviconType = row["iconType"] as? Int,
+           let type = IconType(rawValue: faviconType) {
+                bookmark.favicon = Favicon(url: faviconURL,
+                                           date: NSDate(timeIntervalSince1970: date),
+                                           type: type)
+        }
+    }
+
+    private class func livemarkFactory(row: SDRow) -> BookmarkItem {
+        let id = row["id"] as! Int
+        let guid = row["guid"] as! String
+        let url = row["siteUri"] as! String
```

Comment:
referencing `siteUri` here causes a crash. There is no `siteUri` in the row columnNames. Changing this to `feedUri` seems to work.